codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -137,145 +208,106 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
-        if (childChannel != null) {
-            final EventLoop eventLoop = childChannel.eventLoop();
-            if (eventLoop.inEventLoop()) {
-                onStreamClosed0(childChannel);
-            } else {
-                eventLoop.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        onStreamClosed0(childChannel);
-                    }
-                });
-            }
+            ChannelFuture future = ctx.channel().eventLoop().register(childChannel);
+            future.addListener(CHILD_CHANNEL_REGISTRATION_LISTENER);
         }
-    }
-
-    private void onStreamClosed0(Http2StreamChannel childChannel) {
-        assert childChannel.eventLoop().inEventLoop();
 
-        childChannel.onStreamClosedFired = true;
-        childChannel.fireChildRead(CLOSE_MESSAGE);
+        assert !childChannel.isWritable();
+        childChannel.incrementOutboundFlowControlWindow(initialOutboundStreamWindow);
+        childChannel.pipeline().fireChannelWritabilityChanged();
+        return childChannel;
     }
 
-    void flushFromStreamChannel() {
-        EventExecutor executor = ctx.executor();
-        if (executor.inEventLoop()) {
-            flush(ctx);
-        } else {
-            Runnable task = flushTask;
-            if (task == null) {
-                task = flushTask = new Runnable() {
-                    @Override
-                    public void run() {
-                        flush(ctx);
-                    }
-                };
-            }
-            executor.execute(task);
-        }
+    // TODO: This is most likely not the best way to expose this, need to think more about it.
+    final Http2StreamChannel newOutboundStream() {
+        return newStreamChannel(newStream());
     }
 
-    void writeFromStreamChannel(Object msg, boolean flush) {
-        writeFromStreamChannel(msg, ctx.newPromise(), flush);
+    private  DefaultHttp2StreamChannel newStreamChannel(Http2FrameStream stream) {
+        DefaultHttp2StreamChannel childChannel = new DefaultHttp2StreamChannel(ctx, stream, firstFrameWriteListener);
+        channels.put(stream, childChannel);
+        return childChannel;
     }
 
-    void writeFromStreamChannel(final Object msg, final ChannelPromise promise, final boolean flush) {
-        EventExecutor executor = ctx.executor();
-        if (executor.inEventLoop()) {
-            writeFromStreamChannel0(msg, flush, promise);
-        } else {
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+        if (cause instanceof Http2FrameStreamException) {
+            Http2FrameStreamException streamException = (Http2FrameStreamException) cause;
+            Http2FrameStream stream = streamException.stream();
+            DefaultHttp2StreamChannel childChannel = channels.get(stream);
+
+            if (childChannel == null)  {
+                // TODO: Should we log this ?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)